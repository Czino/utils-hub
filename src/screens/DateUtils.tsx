import dayjs from 'dayjs'
import LocalizedFormat from 'dayjs/plugin/localizedFormat'
import { useState } from 'react'
import { FiArrowRight } from 'react-icons/fi'
import { Headline } from '../components/Headline'
import { Input } from '../components/Input'
import { Select } from '../components/Select'
import en from '../i18n/en'
import { i18n } from '../i18n/i18n'
import { ScreenWithSideNavigation } from '../templates/ScreenWithSideNavigation'
import allLocales from '../utils/locale/allLocales'
import { keys } from '../utils/object/keys'
const locales = keys(allLocales)
const FORMATS = ['L', 'LL', 'LLL', 'LLLL']
dayjs.extend(LocalizedFormat)

export const DateUtils = () => {
  const [date, setDate] = useState(dayjs().format('YYYY-MM-DD hh:mm'))
  const [locale, setLocale] = useState('en')
  const localizedDate = dayjs(date).locale(locale)
  return (
    <ScreenWithSideNavigation>
      <Headline>{i18n(en.date.title)}</Headline>
      <p>{i18n(en.date.description)}</p>
      <div className="grid grid-cols-3 gap-4 items-center md:grid-cols-9">
        <Input
          className="col-span-3"
          type="datetime-local"
          aria-label={i18n(en.form.date)}
          value={date}
          onChange={(e) => setDate(e.currentTarget.value)}
        />
        <Select
          className="col-span-3 md:col-span-2"
          aria-label={i18n(en.form.locale)}
          value={locale}
          onChange={(value) => setLocale(value.replace('_', '-'))}
        >
          {locales.map((l) => (
            <option key={l} value={l}>
              {allLocales[l].name} ({i18n(en.language[l])})
            </option>
          ))}
        </Select>
        <div className="grid col-span-3 justify-center items-center self-center h-full md:col-span-1">
          <FiArrowRight />
        </div>
        <div className="col-span-3">
          {localizedDate.isValid() && (
            <ul>
              <li>{localizedDate.toISOString()}</li>

              {FORMATS.map((format) => (
                <li key={format}>{localizedDate.format(format)}</li>
              ))}
            </ul>
          )}
        </div>
      </div>
    </ScreenWithSideNavigation>
  )
}
