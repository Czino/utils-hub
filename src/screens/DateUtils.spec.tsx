import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { toMatchDiffSnapshot } from 'snapshot-diff'
import { DateUtils } from './DateUtils'
expect.extend({ toMatchDiffSnapshot })

jest.useFakeTimers({ now: new Date('2022-05-03T10:59Z') })

describe('DateUtils', () => {
  const dateLabel = 'input date'
  const localeLabel = 'select locale'
  const value = '2023-01-04 19:32'
  const base = render(<DateUtils />).asFragment()

  afterEach(() => {
    jest.runOnlyPendingTimers()
    jest.useRealTimers()
  })
  it('renders default', () => {
    expect(base).toMatchSnapshot()
  })
  it('updates date and displays in given format', async () => {
    const expected = '2023-01-04T19:32'
    const { asFragment, getByLabelText } = render(<DateUtils />)
    const $date = getByLabelText(dateLabel)
    await userEvent.clear($date)
    await userEvent.type($date, value)
    expect($date).toHaveValue(expected)
    expect(asFragment()).toMatchDiffSnapshot(base)
  })
  it('updates locale and displays localized format', async () => {
    const { asFragment, getByLabelText } = render(<DateUtils />)
    const $date = getByLabelText(dateLabel)
    const $locale = getByLabelText(localeLabel)
    await userEvent.clear($date)
    await userEvent.type($date, value)
    await userEvent.clear($locale)
    await userEvent.type($locale, 'de')
    await userEvent.click(getByLabelText('option de (German)'))
    expect($locale).toHaveValue('de (German)')
    expect(asFragment()).toMatchDiffSnapshot(base)
  })
})
