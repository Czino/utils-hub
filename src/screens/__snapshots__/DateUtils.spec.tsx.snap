// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateUtils renders default 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row gap-4 text-2xl color-black"
  >
    <div
      class="fixed inset-0 transition-[opacity] bg-black pointer-events-none opacity-0"
    />
    <nav
      class="bg-white fixed z-10 p-4 transition-[top] overflow-auto pb-32 w-full h-full top-full md:h-screen md:relative md:left-0 md:w-[240px]"
    >
      <ul>
        <li
          class="font-bold"
        >
          <h1>
            <a
              href="/"
            >
              UtilityHub
            </a>
          </h1>
        </li>
        <li>
          <h2
            class="mt-4 font-bold"
          >
            Time
          </h2>
        </li>
        <li>
          <a
            class=""
            href="/utils/date"
          >
            Dates
          </a>
        </li>
        <li>
          <h2
            class="mt-4 font-bold"
          >
            Numbers
          </h2>
        </li>
        <li>
          <a
            class=""
            href="/utils/fractions"
          >
            Fractions
          </a>
        </li>
        <li>
          <a
            class=""
            href="/utils/percentages"
          >
            Percentages 
          </a>
        </li>
        <li>
          <a
            class=""
            href="/utils/units"
          >
            Units
          </a>
        </li>
        <li>
          <h2
            class="mt-4 font-bold"
          >
            Text
          </h2>
        </li>
        <li>
          <a
            class=""
            href="/utils/text"
          >
            Text
          </a>
        </li>
        <li>
          <a
            class=""
            href="/utils/list"
          >
            Lists
          </a>
        </li>
        <li>
          <a
            class=""
            href="/utils/encoding"
          >
            Encoding
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="flex fixed right-0 bottom-0 z-20 justify-end p-4 w-full bg-white  md:hidden  border shadow-2xl border-top-1"
      role="button"
    >
      <svg
        aria-label="open navigation"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="3"
          x2="21"
          y1="12"
          y2="12"
        />
        <line
          x1="3"
          x2="21"
          y1="6"
          y2="6"
        />
        <line
          x1="3"
          x2="21"
          y1="18"
          y2="18"
        />
      </svg>
    </div>
    <div
      class="p-4 space-y-4 w-full"
    >
      <h1
        class="text-5xl text-success-2 font-bold "
      >
        Date Format Converter
      </h1>
      <p>
        Whether you're a developer wrestling with date formats between systems, a professional managing international projects, or a student dealing with diverse date formats in research, our tool is designed to cater to your needs. Instantly convert dates between different formats, including ISO, US, European, and more, with ease and precision. Save time and eliminate errors in your work with our user-friendly date conversion tool.
      </p>
      <div
        class="grid grid-cols-3 gap-4 items-center md:grid-cols-9"
      >
        <input
          aria-label="input date"
          class="w-full text-xs md:text-xl appearance-none bg-transparent text-dark p-2 border border-dark rounded col-span-3"
          type="datetime-local"
          value="2022-05-03 12:59"
        />
        <div
          class="relative col-span-3 md:col-span-2"
        >
          <input
            aria-label="select locale"
            class="p-2 w-full text-xs bg-transparent rounded border md:text-xl border-dark"
            type="text"
            value="en (English)"
          />
          <svg
            class="absolute right-1 top-1/2 w-6 h-6 transform -translate-y-1/2 pointer-events-none"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </div>
        <div
          class="grid col-span-3 justify-center items-center self-center h-full md:col-span-1"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="5"
              x2="19"
              y1="12"
              y2="12"
            />
            <polyline
              points="12 5 19 12 12 19"
            />
          </svg>
        </div>
        <div
          class="col-span-3"
        >
          <ul>
            <li>
              2022-05-03T10:59:00.000Z
            </li>
            <li>
              05/03/2022
            </li>
            <li>
              May 3, 2022
            </li>
            <li>
              May 3, 2022 12:59 PM
            </li>
            <li>
              Tuesday, May 3, 2022 12:59 PM
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="w-[280px] hidden md:block"
    />
  </div>
</DocumentFragment>
`;

exports[`DateUtils updates date and displays in given format 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -151,11 +151,11 @@
        >
          <input
            aria-label="input date"
            class="w-full text-xs md:text-xl appearance-none bg-transparent text-dark p-2 border border-dark rounded col-span-3"
            type="datetime-local"
-           value="2023-01-04T19:32"
+           value="2022-05-03 12:59"
          />
          <div
            class="relative col-span-3 md:col-span-2"
          >
            <input
@@ -209,23 +209,23 @@
          <div
            class="col-span-3"
          >
            <ul>
              <li>
-               2023-01-04T18:32:00.000Z
+               2022-05-03T10:59:00.000Z
              </li>
              <li>
-               01/04/2023
+               05/03/2022
              </li>
              <li>
-               January 4, 2023
+               May 3, 2022
              </li>
              <li>
-               January 4, 2023 7:32 PM
+               May 3, 2022 12:59 PM
              </li>
              <li>
-               Wednesday, January 4, 2023 7:32 PM
+               Tuesday, May 3, 2022 12:59 PM
              </li>
            </ul>
          </div>
        </div>
      </div>"
`;

exports[`DateUtils updates locale and displays localized format 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -151,20 +151,20 @@
        >
          <input
            aria-label="input date"
            class="w-full text-xs md:text-xl appearance-none bg-transparent text-dark p-2 border border-dark rounded col-span-3"
            type="datetime-local"
-           value="2023-01-04T19:32"
+           value="2022-05-03 12:59"
          />
          <div
            class="relative col-span-3 md:col-span-2"
          >
            <input
              aria-label="select locale"
              class="p-2 w-full text-xs bg-transparent rounded border md:text-xl border-dark"
              type="text"
-             value="de (German)"
+             value="en (English)"
            />
            <svg
              class="absolute right-1 top-1/2 w-6 h-6 transform -translate-y-1/2 pointer-events-none"
              fill="none"
              height="1em"
@@ -209,23 +209,23 @@
          <div
            class="col-span-3"
          >
            <ul>
              <li>
-               2023-01-04T18:32:00.000Z
+               2022-05-03T10:59:00.000Z
              </li>
              <li>
-               04.01.2023
+               05/03/2022
              </li>
              <li>
-               4. Januar 2023
+               May 3, 2022
              </li>
              <li>
-               4. Januar 2023 19:32
+               May 3, 2022 12:59 PM
              </li>
              <li>
-               Mittwoch, 4. Januar 2023 19:32
+               Tuesday, May 3, 2022 12:59 PM
              </li>
            </ul>
          </div>
        </div>
      </div>"
`;
